scheds/rust/scx_rustland/src/bpf/main.bpf.c

```diff
@@ -276,16 +280,23 @@ static void dispatch_local(struct task_struct *p, u64 enq_flags)
 }
 
 /*
- * Dispatch a task on a target per-CPU FIFO.
+ * Dispatch a task on a target CPU.
+ *
+ * If it's not possible to dispatch on the selected CPU, try to re-use the
+ * previously used one, if that one is also not usable dispatch to the global
+ * DSQ.
  */
-static void dispatch_on_cpu(struct task_struct *p, s32 cpu, u64 enq_flags)
+static void dispatch_task(struct task_struct *p, s32 cpu, u64 enq_flags)
 {
-	/*
-	 * If it's not possible to dispatch on the selected CPU, re-use the
-	 * previously used one.
-	 */
-	if (!bpf_cpumask_test_cpu(cpu, p->cpus_ptr))
+	if (!bpf_cpumask_test_cpu(cpu, p->cpus_ptr)) {
 		cpu = scx_bpf_task_cpu(p);
+		if (!bpf_cpumask_test_cpu(cpu, p->cpus_ptr)) {
+			dbg_msg("%s: pid=%d global", __func__, p->pid);
+			scx_bpf_dispatch(p, SCX_DSQ_GLOBAL, slice_ns,
+					 enq_flags);
```

Comment:
I wonder if this should eventually be an error condition in the scheduler? Unfortunately we don't yet have an ergonomic API for looking at a task's cpumask in user space so it might be a bit premature.

Also, on the other hand, maybe it would make the scheduler unnecessarily brittle in that a task could exit or something while it's being dispatched from user space. So this fallback is probably fine, but maybe we should add a comment explaining why this could happen?

Reply:
> I wonder if this should eventually be an error condition in the scheduler? Unfortunately we don't yet have an ergonomic API for looking at a task's cpumask in user space so it might be a bit premature.

Oh yes! Having an API to access cpumask would be super useful... we could use that in the scheduler to avoid selecting dispatching tasks on CPUs that can't be used.

Right now we could report that as a warning maybe? I don't think we should error the scheduler, it doesn't seem to be a too critical condition, considering that in the worst case scenario we can simply use the global DSQ and everything will work just fine.

Moreover even if we have access to the cpumask in the future we may still do something similar, unless we also provide some mechanisms to lock the cpumask.

> 
> Also, on the other hand, maybe it would make the scheduler unnecessarily brittle in that a task could exit or something while it's being dispatched from user space. So this fallback is probably fine, but maybe we should add a comment explaining why this could happen?

Absolutely, I'll add a comment. Maybe even introduce an extra counter, like `nr_dispatch_failures` or something similar?

Reply:
>Right now we could report that as a warning maybe? I don't think we should error the scheduler, it doesn't seem to be a too critical condition, considering that in the worst case scenario we can simply use the global DSQ and everything will work just fine.

Yeah, agreed. And I don't think it necessarily indicates a bad / corrupted scheduler or anything anyways given that everything is async, so this is inherently a bit racy.

>Absolutely, I'll add a comment. Maybe even introduce an extra counter, like nr_dispatch_failures or something similar?

SGTM!